docs/tvtime-scanner.1
.\" Man page for tvtime-scanner
.\" Copyright (c) 2003, 2004  Billy Biggs
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or (at
.\" your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH tvtime-scanner 1 "July 2004" "tvtime 0.9.13"

.SH NAME
tvtime\-scanner \- an all-band scanner for
.B tvtime
to help users of European cable providers that do not use standard frequencies 

.SH SYNOPSIS

.B tvtime\-scanner
[
.IR OPTION
] ...

.SH DESCRIPTION

.B tvtime\-scanner
allows a user to scan all frequencies for stations.  This is useful for European
cable provides such as
.B UPC
or
.B Casema
which do not use standard frequency tables for their channels.
.B tvtime-scanner
checks for frequencies across the full band and writes it to your
stationlist.xml file in a table called "Custom".  This can be selected by
running "tvtime --frequencies=custom" or by selecting the Custom frequency
table within the OSD menu inside
.B tvtime.

.SH OPTIONS

.TP
.B \-h, \-\-help
Show usage information.

.TP
.B \-F, \-\-configfile=FILE
Additional config file to load settings from.

.TP
.B \-d, \-\-device=NAME
video4linux device to use (defaults to
.IR /dev/video0 ).

.TP
.B \-i, \-\-input=INPUTNUM
video4linux input number (defaults to
.IR 0 ).
Cards commonly have a bunch
of sources, for example, on my WinTV card,
.I source 0
is my tuner and
.I source 1
is the composite input.  The scanner requires using an input
that has a tuner.

.TP
.B \-n, \-\-norm=NORM
The mode to set the tuner to (defaults to
.IR NTSC ).
Valid modes are:
.TS
nokeep tab (@);
l l.
\(bu@NTSC
\(bu@PAL
\(bu@SECAM
\(bu@PAL\-NC
\(bu@PAL\-M
\(bu@PAL\-N
\(bu@NTSC\-JP
.TE

.SH AUTHORS

Billy Biggs.

.SH "SEE ALSO"

.BR tvtime (1),
.BR tvtime-configure (1),
.IR tvtime.xml (5),


NEWS

For news and updates on tvtime, please visit our website at:

  http://tvtime.net/


News for 1.0

  *  By default, tvtime now assumes that pixels are square.  This
     avoids problems with many misconfigured X servers.

News for 0.99

  *  Test release for 1.0.
  *  Major bugfixes and cleanup.

News for 0.9.15

  *  Compile fixes.

News for 0.9.14

  *  Improved the usability of the menu system when using a mouse.
  *  Fixed some problems with Xinerama and fullscreen.
  *  Streamlined the default set of key bindings.
  *  Show an icon in modern window list applications.
  *  Fixed colour problems for DVB driver users.
  *  Changed the default process priority to -10 instead of -19.
  *  Fixed compile problems with new gcc versions.

News for 0.9.13

  *  Added support for changing the default language read from XMLTV.
  *  Added support for PAL-BG/DK audio mode switching.
  *  Added an audio boost setting for controlling capture card volume.
  *  Fixed a bunch of critical bugs.
  *  Removed the SDL, MGA, and DirectFB output drivers.
  *  Removed the unfinished MPEG2 support and the unreleased recording
     application 'rvr'.

News for 0.9.12

  *  Minor bugfixes from 0.9.11.

News for 0.9.11

  *  Fixed gettext to initialize properly.
  *  Improved our XMLTV on-screen-display.
  *  Display the upcoming show name from XMLTV.
  *  Changed fifodir back to /tmp.
  *  Lots of documentation improvements.
  *  Fixed lots of bugs.

News for 0.9.10

  *  Added program listings using XMLTV.  Use with --xmltv=filename.
  *  Added commands for chroma kill, sleep timer, and enabling or
     disabling signal detection.
  *  Added the matte mode setting to the menu.
  *  Added a DirectFB output layer by Ville Syrjala.
  *  Significantly lowered the size of the tvtime helper executables.
  *  Various menu system cleanups.
  *  Fixed a ton of bugs.

News for 0.9.9

  *  Added a full menu system for configuring tvtime.
  *  Simplified deinterlacer names and provide online help for them.
  *  Unified OSD appearance across all television standards.
  *  Support for saving picture settings both globally and per-channel.
  *  Finetune settings saved per-channel.
  *  Support for the V4L2 DVB driver and bttv9 for PAL-60 support.
  *  Added a channel scanner application 'tvtime-scanner' that outputs
     to the 'Custom' frequency table.
  *  Native support for UYVY input from the rivatv driver.
  *  Removed lirc support in favour of using tvtime-command.
  *  Fixed remaining screensaver, Xinerama and fullscreen issues.
  *  Many other new features and bugfixes.

News for 0.9.8.5

  *  Some root safety improvements.
  *  Added support for long-format command line arguments.
  *  Added config file options to set startup window position.
  *  Compile fixes for non-C99 compilers and older DirectFB versions.

News for 0.9.8.4

  *  Fixed to work with metacity in RH 8.0.

News for 0.9.8.3

  *  Support for norm switching between "compatible" norms for areas
     which receive both SECAM and PAL format stations.
  *  Added support for the mga_vid driver for the Matrox G200/G400
     series cards from mplayer.
  *  Semi-experimental of the X code for better fullscreen support under
     many window managers.  Testing appreciated.
  *  Support for changing the mixer device and channel (see 'man tvtime'
     and the config file documentation).
  *  Slightly less partial experimental UTF-8 support. Now works with
     any simple left-to-right scripts which are in FreeSansBold.ttf .

News for 0.9.8.2

  *  Fixed to compile on older gcc versions (blame pv2b).

News for 0.9.8.1

  *  Alot of minor bugfixes.
  *  Partial EXPERIMENTAL UTF-8 support. Only works with UTF-8
     characters that are also in ISO-8859-1 (Latin-1).

News for 0.9.8

  Users of tvtime 0.9.7:
    The default key bindings and configuration file format have
    significantly changed in this release of tvtime.  Please see the
    documentation for details on using tvtime 0.9.8.

    Also, some command line options have changed or become unavailable
    (no more -w, use -H instead).  Please consult the tvtime man page.

  *  New XML config file format for both the tvtime configuration,
     and the station list.
  *  tvtime now has two distinct half-framerate modes: top-field-first
     frames, and bottom-field-first frames.  This helps with manually
     deinterlacing 2-2 pulldown or progressive content.
  *  Added support for preset 'modes', which can set the deinterlacer,
     window size, fullscreen setting and framerate mode.
  *  Added support for custom frequencies in the stationlist.xml file,
     and a runtime command to add a new station from the current
     fine-tuning settings.
  *  Added some DScaler DLLs and a wine layer to load these deinterlacer
     plugins directly without porting to Linux.  This gives us the
     following new deinterlacers:
       GreedyH        DScaler: Greedy - High Motion
       Greedy2Frame   DScaler: Greedy - 2-frame
       TwoFrame       DScaler: TwoFrame
       TomsMoComp     DScaler: TomsMoComp
       VideoBob       DScaler: Video Bob
       VideoWeave     DScaler: Video Weave
     We had our own ports of videobob, twoframe, and greedy2frame,
     but they were found to be buggy.  Our ports are currently disabled.
  *  Added the ffmpeg vertical deinterlacer filter.
  *  Support for multiple audio channels (stereo, mono, lang1, lang2).
  *  Included a bunch of fixes for PAL-M support.
  *  Added a slave mode for freevo integration.
  *  Various speedups, quality improvements, and bugfixes.

News for 0.9.7

  *  Moved config files into a new config file directory ~/.tvtime
  *  Reworked the frequency tables, merging europe-cable, europe-west,
     europe-east all into 'europe'.  Added the 'russia' frequency table.
  *  Added a channel scanner.  F10 walks through the channel list and
     disables inactive channels.
  *  Added the 'XVideo Bob' deinterlacer filter.
  *  Added lirc support.

src/tvtime-scanner.c
/**
 * Copyright (c) 2003 Billy Biggs <vektor@dumbterm.net>.
 *
 * Based heavily on 'scantv.c' from xawtv,
 *   (c) 2000-2002 Gerd Knorr <kraxel@goldbach.in-berlin.de>
 * See http://bytesex.org/xawtv/
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */


#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include "videoinput.h"
#include "tvtimeconf.h"
#include "station.h"
#include "utils.h"

int main( int argc, char **argv )
{
    config_t *cfg = config_new();
    station_mgr_t *stationmgr = 0;
    videoinput_t *vidin;
    int fi, on, tuned;
    int f, f1, f2, fc;
    int verbose, norm;
    int curstation = 1;

    if( !cfg ) {
        fprintf( stderr, "tvtime-scanner: Can't initialize tvtime configuration, exiting.\n" );
        return 1;
    }

    if( !config_parse_tvtime_scanner_command_line( cfg, argc, argv ) ) {
        config_delete( cfg );
        return 1;
    }

    verbose = config_get_verbose( cfg );

    if( !strcasecmp( config_get_v4l_norm( cfg ), "pal" ) ) {
        norm = VIDEOINPUT_PAL;
    } else if( !strcasecmp( config_get_v4l_norm( cfg ), "secam" ) ) {
        norm = VIDEOINPUT_SECAM;
    } else if( !strcasecmp( config_get_v4l_norm( cfg ), "pal-nc" ) ) {
        norm = VIDEOINPUT_PAL_NC;
    } else if( !strcasecmp( config_get_v4l_norm( cfg ), "pal-m" ) ) {
        norm = VIDEOINPUT_PAL_M;
    } else if( !strcasecmp( config_get_v4l_norm( cfg ), "pal-n" ) ) {
        norm = VIDEOINPUT_PAL_N;
    } else if( !strcasecmp( config_get_v4l_norm( cfg ), "ntsc-jp" ) ) {
        norm = VIDEOINPUT_NTSC_JP;
    } else if( !strcasecmp( config_get_v4l_norm( cfg ), "pal-60" ) ) {
        norm = VIDEOINPUT_PAL_60;
    } else {
        /* Only allow NTSC otherwise. */
        norm = VIDEOINPUT_NTSC;
    }

    fprintf( stderr, "tvtime-scanner: Scanning using TV standard '%s'\n",
             videoinput_get_norm_name( norm ) );

    stationmgr = station_new( videoinput_get_norm_name( norm ), "Custom", 0, verbose );
    if( !stationmgr ) {
        fprintf( stderr, "tvtime-scanner: Can't create station manager (no memory?), exiting.\n" );
        config_delete( cfg );
        return 1;
    }

    vidin = videoinput_new( config_get_v4l_device( cfg ), 
                            config_get_inputwidth( cfg ), 
                            norm, verbose );
    if( !vidin ) {
        fprintf( stderr, "tvtime-scanner: Can't open video4linux device '%s'.\n",
                config_get_v4l_device( cfg ) );
        station_delete( stationmgr );
        config_delete( cfg );
        return 1;
    } else {
        videoinput_set_input_num( vidin, config_get_inputnum( cfg ) );
    }

    if( !videoinput_has_tuner( vidin ) ) {
        fprintf( stderr, "tvtime-scanner: No tuner found on input %d.\n"
                 "tvtime-scanner: If you have a tuner, select a different input using --input=<num>.\n",
                 config_get_inputnum( cfg ) );
        videoinput_delete( vidin );
        station_delete( stationmgr );
        config_delete( cfg );
        return 1;
    }

    /* Scan freqnencies */
    fprintf( stderr, "tvtime-scanner: Scanning from %6.2fMHz to %6.2fMHz.\n",
             44.0, 958.0 );
    on = 0;
    fc = 0;
    f1 = 0;
    f2 = 0;
    fi = -1;

    for( f = 44*16; f <= 958*16; f += 4 ) {
        char stationmhz[ 128 ];

        videoinput_set_tuner_freq( vidin, (f * 1000) / 16 );
        fprintf( stderr, "tvtime-scanner: Checking %6.2fMHz:", ((double) f) / 16.0 );
        usleep( 200000 ); /* 0.2 sec */
        tuned = videoinput_freq_present( vidin );

        /* state machine */
        if( 0 == on && 0 == tuned ) {
            fprintf( stderr, "  - No signal      \r" );
            continue;
        }
        if( 0 == on && 0 != tuned ) {
            fprintf( stderr, "  + Signal detected\r" );
            f1 = f;
            /* if( i != chancount ) { fi = i; fc = f; } */
            on = 1;
            continue;
        }
        if( 0 != on && 0 != tuned ) {
            fprintf( stderr, "  * Signal detected\r" );
            /* if( i != chancount ) { fi = i; fc = f; } */
            continue;
        }
        /* if (on != 0 && 0 == tuned)  --  found one, read name from vbi */
        f2 = f;
        if( 0 == fc ) {
            fc = (f1 + f2) / 2;
            /* Round to the nearest .25MHz */
            fc = ((fc + 2)/4)*4;
        }

        fprintf( stderr, "\rtvtime-scanner: Found a channel at %6.2fMHz (%.2f-%.2fMHz), adding to stationlist.\n",
                 ((double) fc) / 16.0, ((double) f1) / 16.0, ((double) f2) / 16.0 );

        sprintf( stationmhz, "%.2fMHz", ((double) fc) / 16.0 );
        station_add( stationmgr, curstation, "Custom", stationmhz, stationmhz );
        station_writeconfig( stationmgr );
        curstation++;

        on = 0;
        fc = 0;
        f1 = 0;
        f2 = 0;
        fi = -1;
    }

    station_delete( stationmgr );
    config_delete( cfg );
    return 0;
}

